check-version.sh

```diff
@@ -10,29 +10,39 @@ trap 'echo "$0: Check failed. Stopping." >&2' ERR
 readonly version_path='VERSION'
 readonly changes_path='doc/source/changes.rst'
 
+function check_status() {
+    git status -s "$@"
+    test -z "$(git status -s "$@")"
+}
+
+function get_latest_tag() {
+    local config_opts
+    printf -v config_opts ' -c versionsort.suffix=-%s' alpha beta pre rc RC
+    # shellcheck disable=SC2086  # Deliberately word-splitting the arguments.
+    git $config_opts tag -l '[0-9]*' --sort=-v:refname | head -n1
+}
+
 echo 'Checking current directory.'
 test "$(cd -- "$(dirname -- "$0")" && pwd)" = "$(pwd)"  # Ugly, but portable.
 
 echo "Checking that $version_path and $changes_path exist and have no uncommitted changes."
 test -f "$version_path"
 test -f "$changes_path"
-git status -s -- "$version_path" "$changes_path"
-test -z "$(git status -s -- "$version_path" "$changes_path")"
+check_status -- "$version_path" "$changes_path"
 
 # This section can be commented out, if absolutely necessary.
 echo 'Checking that ALL changes are committed.'
-git status -s --ignore-submodules
-test -z "$(git status -s --ignore-submodules)"
+check_status --ignore-submodules
 
-version_version="$(cat "$version_path")"
+version_version="$(<"$version_path")"
 changes_version="$(awk '/^[0-9]/ {print $0; exit}' "$changes_path")"
-config_opts="$(printf ' -c versionsort.suffix=-%s' alpha beta pre rc RC)"
-latest_tag="$(git $config_opts tag -l '[0-9]*' --sort=-v:refname | head -n1)"
+latest_tag="$(get_latest_tag)"
 head_sha="$(git rev-parse HEAD)"
 latest_tag_sha="$(git rev-parse "${latest_tag}^{commit}")"
 
 # Display a table of all the current version, tag, and HEAD commit information.
-echo $'\nThe VERSION must be the same in all locations, and so must the HEAD and tag SHA'
+echo
```

Comment:
I did it for spacing, and to to take [fuller advantage of `echo` to have fewer `\n` sequences](https://github.com/gitpython-developers/GitPython/pull/1661#issuecomment-1747179585). Although it's true that `$'` `'` is not POSIX and shouldn't be used in a `#!/bin/sh` script, this script is a `bash` script and `$'` `'` is a available in `bash` on all systems. (`ksh`, `zsh`, and probably some other shells also have `$'` `'`, but not all POSIX-compatible shells have it.)

Regarding spacing, the table is currently formatted using `printf` and this has the advantage that its own output spacing can be adjusted by, for example, changing `14` to `15`. But I originally used `printf` for it because I was preferring `printf` to `echo`, because originally I was [doing this inline in `Makefile`](https://github.com/gitpython-developers/GitPython/commit/6495d84142b60ba81a5b4268a0dfc0785c22d60a#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52), and a `Makefile` runs commands using `/bin/sh` (unless `SHELL` is assigned in the `Makefile` itself). This was to follow the practice of avoiding `echo` in portable `sh` scripts, [at least when displaying data read from files](https://unix.stackexchange.com/a/65819/11938).

But that is not necessary in a `#!/bin/bash` script, where we know how `echo` works and where `echo` doesn't expand escape sequences by default. So it would actually be okay, at this point, to go back to displaying the table this way:

```bash
echo
echo 'The VERSION must be the same in all locations, and so must the HEAD and tag SHA'
echo "VERSION file   = $version_version"
echo "changes.rst    = $changes_version"
echo "Latest tag     = $latest_tag"
echo "HEAD SHA       = $head_sha"
echo "Latest tag SHA = $latest_tag_sha"
```

This is arguably more readable, and arguably conveys spacing the best. I'd be happy to change it to this if you like.

(Another option could be to go in the opposite direction and make this `check-version.sh` script, and maybe the `build-release.sh` script as well, `#!/bin/sh` script, which is feasible.)

Reply:
> This is arguably more readable, and arguably conveys spacing the best. I'd be happy to change it to this if you like.
> (Another option could be to go in the opposite direction and make this `check-version.sh` script, and maybe the `build-release.sh` script as well, a `#!/bin/sh` script, which is feasible.)

Since `trap` is supported in POSIX shells and thus `sh`, making these scripts more portable seems valuable, even though I think that as the project currently is setup there isn't any need as it's only me using them.
Alternatives involve running these on CI which also supports `bash` everywhere (at least so it seems).

If you think there is benefits to portability assuming that one day CI can perform trusted publishes, I think it's fair to adjust for portability next time you touch them. Otherwise, it would be just as fair to change `printf` with `echo` as shown above. It's perfectly optional as well.

Reply:
The desired interaction of `trap` and `set -e` may be achievable in `sh`. Even if not, we don't have to use `trap`; it can be replaced even without making the control-flow logic more complicated by, for example, having the script accept `-q` to not append the failure message, and otherwise invoke itself with `-q`, check the result, print the message if it failed, and exit with the original failing exit code. (There is also [an argument to be made](https://mywiki.wooledge.org/BashFAQ/105) that `set -e` shouldn't be relied on at all.)

There are not many systems where `bash` doesn't exist and can't be made available. However, due to the problem described in 729372f--which only happens when MinGW `make` is being used on a native Windows system that has also WSL installed, and does not happen outside `make` nor on other systems--the scripts' `bash` hashbangs are written as `#!/bin/bash` rather than `#!/usr/bin/env bash`, and thus assume `bash` is in `/bin`.

Making them `#!/bin/sh` scripts would also fix that. It is even safer to assume `sh` is in `bin` than to assume `env` is in `/usr/bin`, so `#!/usr/bin/env sh` is rarely used. (Anyway, there is no corresponding WSL-related `sh.exe` in `System32`.) So, when combined with the other minor reasons for using `#!/bin/sh`, that is probably worth doing.

As you've suggested, next time I work on those scripts I'll look into making them POSIX `sh` scripts. However, this would be unrelated to facilitating trusted publishing. CI does support `bash` everywhere, at least on runners hosted by GitHub Actions.